pkg/internalregistry/controller.go

```diff
@@ -0,0 +1,333 @@
+	"k8s.io/client-go/kubernetes"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/credentialprovider"
+)
+
+const (
+	controllerName                        = string(openshiftcontrolplanev1.OpenShiftServiceAccountPullSecretsController)
+	InternalRegistryImagePullSecretRefKey = "openshift.io/internal-registry-pull-secret-ref"
+	ExpirationCheckPeriod                 = 10 * time.Minute
+)
+
+type ImagePullSecretsController interface {
+	factory.Controller
+}
+
+type imagePullSecretsController struct {
+	factory.Controller
+	kubeClient           kubernetes.Interface
+	secretLister         corev1listers.SecretLister
+	serviceAccountLister corev1listers.ServiceAccountLister
+	serviceLister        corev1listers.ServiceLister
+	recorder             events.Recorder
+
+	additionalRegistryURLs []string
+}
+
+func NewImagePullSecretsController(
+	kubeClient kubernetes.Interface,
+	serviceAccountInformerFactory corev1informers.ServiceAccountInformer,
+	secretInformerFactory corev1informers.SecretInformer,
+	servicesInformerFactory corev1informers.ServiceInformer,
+	additionalRegistryURLs []string,
+	recorder events.Recorder,
+) *imagePullSecretsController {
+	c := &imagePullSecretsController{
+		kubeClient:             kubeClient,
+		secretLister:           secretInformerFactory.Lister(),
+		serviceAccountLister:   serviceAccountInformerFactory.Lister(),
+		serviceLister:          servicesInformerFactory.Lister(),
+		additionalRegistryURLs: additionalRegistryURLs,
+		recorder:               recorder,
+	}
+	triggers := []factory.Informer{
+		secretInformerFactory.Informer(),
+		serviceAccountInformerFactory.Informer(),
+	}
+	c.Controller = factory.New().
+		WithInformers(triggers...).
+		WithFilteredEventsInformers(servicesFilter, servicesInformerFactory.Informer()).
+		ResyncEvery(ExpirationCheckPeriod).
+		WithSync(c.sync).
+		ToController(controllerName, recorder)
+	return c
+}
+
+func (c *imagePullSecretsController) sync(ctx context.Context, syncContext factory.SyncContext) error {
+	serviceAccounts, err := c.serviceAccountLister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	var errs []error
+	for _, serviceAccount := range serviceAccounts {
+		errs = append(errs, c.syncServiceAccount(ctx, serviceAccount))
+	}
+	return errors.Join(errs...)
+}
+
+func (c *imagePullSecretsController) syncServiceAccount(ctx context.Context, serviceAccount *corev1.ServiceAccount) error {
+	var err error
+	// ensure image pull secret name annotation
+	serviceAccount, err = c.syncGeneratedImagePullSecretName(ctx, serviceAccount)
+	if err != nil {
+		return err
+	}
+
+	// get generated image pull secret for the service account
+	imagePullSecret, err := c.imagePullSecretForServiceAccount(ctx, serviceAccount)
+	if err != nil {
+		return err
+	}
+
+	// update registry authentication in the generated image pull secret
+	err = c.syncImagePullSecret(ctx, imagePullSecret, serviceAccount)
```

Comment:
service accounts can have more than one imagepullsecret, this appears to be the only callsite. Perhaps a controller that inspects image pull secrets would be robust